### PR TITLE
Add ADR for proxying S3 file uploads through CloudFront

### PR DESCRIPTION
**Pull request by suzannehamilton**

### CODE REVIEW

File: architecture-decision-records/0022-cloudfront-upload-proxy.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
+# 22. Proxy S3 uploads through CloudFront
+
+**Date:** 2020-09-17
```
Comment (TomJKing) [inline architecture-decision-records/0022-cloudfront-upload-proxy.md:3]: Incorrect year. Should be "2021"

File: architecture-decision-records/0022-cloudfront-upload-proxy.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+- `sts.amazonaws.com`: Exchange the Cognito token for a long-lived STS token
+- `tdr-upload-files-dirty-<environment>.s3.eu-west-2.amazonaws.com`: Upload the files
+
+The sequence of requests looks like this (simplified slightly - CORS requests and multiple requests to the auth server
+and S3 are omitted for clarity):
+
+![Sequence diagram showing current upload process](images/upload-sequence-diagram-cognito.png)
+
+Some TDR users are on restricted networks, and the upload step completely fails. From what we could see on a video call
+with one user, the network calls to TDR domains are succeeding but the ones to `amazonaws.com` are being blocked.
```
Comment (TomJKing) [inline architecture-decision-records/0022-cloudfront-upload-proxy.md:20]: Maybe reword slightly to: "... calls to TDR domains succeed but the ones to ..."

File: architecture-decision-records/0022-cloudfront-upload-proxy.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+way to link a Keycloak user ID to a Cognito ID. Cognito will only generate a user ID when given a valid token for the
+Keycloak user, which requires the user to authenticate themselves, so an admin user cannot convert an arbitrary Keycloak
+ID to a Cognito ID or vice-versa.
+
+Using just the Keycloak token would simplify the workflow. If we use CloudFront, we could configure an [edge function]
+to validate the token and authorise the request. The downside is that it's not obviously compatible with the AWS
+JavaScript SDK. The SDK sets an `Authorization` header based on AWS credentials. It's not clear how you can add a custom
+header based on the Keycloak token. So if we wanted to use just the Keycloak tokens to authorise the S3 upload, we would
+have to write more client-side JavaScript to handle multipart uploads to replace the call to the SDK. It would still be
+worth considering this option if the assumptions change in future, though.
```
Comment (TomJKing) [inline architecture-decision-records/0022-cloudfront-upload-proxy.md:74]: Maybe delete the "..., though" from the end of the sentence. Not sure its needed.

**Review by TomJKing — COMMENTED** [review]
Looks good. Couple of minor editorial suggestions. Feel free to ignore. Find the sequence diagrams especially helpful in explaining the old and new processes.

Probably best if @MancunianSam takes a look through this as well

**Comment by suzannehamilton** [comment]
@MancunianSam @Dclipsham Do either of you have any feedback on this one?

File: architecture-decision-records/0022-cloudfront-upload-proxy.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+to the AWS S3 upload bucket. This means that there are several HTTP requests from the user's browser to AWS domains:
+
+- `cognito-identity.eu-west-2.amazonaws.com`: Exchange a TDR access token for a Cognito token
+- `sts.amazonaws.com`: Exchange the Cognito token for a long-lived STS token
+- `tdr-upload-files-dirty-<environment>.s3.eu-west-2.amazonaws.com`: Upload the files
+
+The sequence of requests looks like this (simplified slightly - CORS requests and multiple requests to the auth server
+and S3 are omitted for clarity):
+
+![Sequence diagram showing current upload process](images/upload-sequence-diagram-cognito.png)
```
Comment (MancunianSam) [inline architecture-decision-records/0022-cloudfront-upload-proxy.md:17]: It's a minor thing which doesn't take away from what you're saying but the diagram says that we exchange the Cognito token for an STS token. We actually call STS to assume a role using the Cognito token and that returns AWS credentials which we then pass to the S3 SDK. So instead of `exchange Cognito for long-lived STS token` it's more `use Cognito token to assume role with STS which returns AWS credentials` and then `check token and S3 key against IAM policy` is `check the S3 key against the IAM policy of the assumed role attached to the credentials`
Or something along those lines anyway. 
This is [the relevant STS call](https://github.com/nationalarchives/tdr-transfer-frontend/blob/master/npm/src/auth/index.ts#L96) which shows us calling `assumeRoleWithWebIdentity`

File: architecture-decision-records/0022-cloudfront-upload-proxy.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+### Proxy options
+
+Whatever option we choose, it will have to solve the problem of sending the data to S3, and authorizing the request so
+that users only have access to S3 objects associated with their own consignments. Another constraint is that we'd like
+to keep using the AWS JavaScript SDK, because the [S3 upload] method automatically handles multipart uploads by chunking
+the file and sending the start and end requests to combine the chunks in S3.
+
+We considered running NGINX in ECS, or writing proxy code and running it in Lambda or ECS. We didn't look very far into
+these options because it means maintaining an extra service. We would also have to proxy the requests to Cognito and
+STS, or find some way to use the AWS SDK to send Keycloak tokens with the upload requests.
```
Comment (MancunianSam) [inline architecture-decision-records/0022-cloudfront-upload-proxy.md:40]: It might be worth mentioning that for the Lambda option, we have a problem with the lambda timeouts, so if a request takes more than the 15 minute max then it would fail. I know it's unlikely with the 5Mb chunking but it's a restriction we didn't have before. 
I also looked briefly at API gateway as a front end to S3 but this has a max upload size of 10Mb and a max request timeout of 30 seconds so that was even less useful. Not sure if it's even worth mentioning :smile:

File: architecture-decision-records/0022-cloudfront-upload-proxy.md (hunk 1 above)
Comment (suzannehamilton) [reply architecture-decision-records/0022-cloudfront-upload-proxy.md]: Whoops. Fixed, thanks.

File: architecture-decision-records/0022-cloudfront-upload-proxy.md (hunk 2 above)
Comment (suzannehamilton) [reply architecture-decision-records/0022-cloudfront-upload-proxy.md]: Done.

File: architecture-decision-records/0022-cloudfront-upload-proxy.md (hunk 3 above)
Comment (suzannehamilton) [reply architecture-decision-records/0022-cloudfront-upload-proxy.md]: 👍

File: architecture-decision-records/0022-cloudfront-upload-proxy.md (hunk 4 above)
Comment (suzannehamilton) [reply architecture-decision-records/0022-cloudfront-upload-proxy.md:18]: Ahh, I'd forgotten the details of that call, thanks. I've reworded mentions of "STS token" to make it clear that STS gives us temporary AWS credentials with a session token. I've updated the first sequence diagram too. Does it make sense?

File: architecture-decision-records/0022-cloudfront-upload-proxy.md (hunk 5 above)
Comment (suzannehamilton) [reply architecture-decision-records/0022-cloudfront-upload-proxy.md:41]: Like you say, the chunking means we wouldn't be hitting timeouts. I think we'd have reduced the chunk size if those Lambdas were returning in more than a few seconds.

So would API Gateway work if you chunked the requests? I guess you'd still have the problem of how to authorise the requests - you'd have the same problem on lines 39-40 about proxying extra requests or adding the Keycloak token to AWS JS SDK requests.

Comment (MancunianSam) [reply architecture-decision-records/0022-cloudfront-upload-proxy.md:41]: API Gateway would have worked with the chunked requests, it's the 30s timeout that would have caused issues I think. Over 30s for a 5Mb chunk isn't unreasonable on a slow connection. We wouldn't have had to bother with Cognito or STS as we could have had a custom authoriser but that would have been unlikely to work nicely with the S3 SDK. I'm happy to leave this bit out anyway as it doesn't really add anything.

File: architecture-decision-records/0022-cloudfront-upload-proxy.md (hunk 4 above)
Comment (MancunianSam) [reply architecture-decision-records/0022-cloudfront-upload-proxy.md:18]: Yep, looks good, thanks.

**Comment by MancunianSam** [comment]
I'm happy with this. @TomJKing @Dclipsham are you happy for this to be approved?

**Comment by TomJKing** [comment]
@MancunianSam @Dclipsham Happy for it to be approved.

**Review by MancunianSam — APPROVED** [review]

File: architecture-decision-records/0022-cloudfront-upload-proxy.md (hunk 5 above)
Comment (suzannehamilton) [reply architecture-decision-records/0022-cloudfront-upload-proxy.md]: I've added a bit about API Gateway, just in case we need to revisit the decision. Does this make sense?

> We considered using API Gateway as a proxy. As with the Lambda or ECS options, we would either have had to proxy the Cognito and STS requests too, or found a way to add the Keycloak token to the AWS JavaScript SDK upload request. API Gateway also has a maximum request time of 30 seconds, which might not be enough for users uploading large files over slow connections. This might have required us to reduce the default chunk size when splitting the file for upload.

**Review by MancunianSam — APPROVED** [review]
